src/vaadin-custom-field.html

```diff
@@ -47,8 +69,199 @@
+            },
+
+            __errorId: String,
+            __labelId: String
           };
         }
+
+        static get observers() {
+          return [
+            '__getActiveErrorId(invalid, errorMessage, __errorId)',
+            '__getActiveLabelId(label, __labelId)',
+            '__updateRequired(required, __inputs)'
+          ];
+        }
+
+        ready() {
+          super.ready();
+
+          this.$.slot.addEventListener('slotchange', e =>
+            this.__setInputs(e.target.assignedNodes().filter(node => node.validate || node.checkValidity)));
+
+          var uniqueId = Vaadin.CustomFieldElement._uniqueId = 1 + Vaadin.CustomFieldElement._uniqueId || 0;
+          this.__errorId = `${this.constructor.is}-error-${uniqueId}`;
+          this.__labelId = `${this.constructor.is}-label-${uniqueId}`;
+        }
+
+        __updateValue(e) {
+          // Stop native change events
+          e && e.stopPropagation();
+
+          this.validate();
+          this.value = this.i18n.formatValue.apply(this, [this.__inputs.map(input => input.value)]);
+          this.dispatchEvent(new CustomEvent('change', {
+            bubbles: true,
+            cancelable: false,
+            detail: {
+              value: this.value
+            }
+          }));
+        }
+
+        __invalidChanged(invalid) {
+          this.__setOrToggleAttribute('aria-invalid', invalid, this);
+        }
+
+        __valueChanged(value) {
+          this.i18n.setValue(value, this.__inputs);
+          this.validate();
+        }
+
+        __setInputs(nodes) {
+          this.__inputs = nodes;
+          this.__inputs.forEach(input => input.addEventListener('change', e => this.__updateValue(e)));
+          this.validate();
+        }
+
+        __updateRequired(required, inputs) {
+          if (!inputs) {
+            return;
+          }
+
+          if (required) {
+            inputs.forEach(input => input.setAttribute('required', ''));
+          } else {
+            inputs.forEach(input => input.removeAttribute('required'));
+          }
+        }
+
+        _labelChanged(label) {
+          if (label !== '' && label != null) {
+            this.setAttribute('has-label', '');
+          } else {
+            this.removeAttribute('has-label');
+          }
+        }
+
+        /**
+         * Returns true if `value` is valid.
+         * `<iron-form>` uses this to check the validity or all its elements.
+         *
+         * @return {boolean} True if the value is valid.
+         */
+        validate() {
+          return !(this.invalid = !this.checkValidity());
+        }
+
+        /**
+         * Returns true if the current inputs values satisfy all constraints (if any)
+         * @returns {boolean}
+         */
+        checkValidity() {
+          if (this.required) {
+            return this.__inputs.every(input => (input.validate || input.checkValidity).call(input));
+          } else {
+            return !this.invalid;
```

Comment:
BTW this line will likely cause `invalid` not to be reset if the previously invalid field becomes valid. Need to test this scenario, we had this issue in combo-box.